mmap-alloc/src/lib.rs

```diff
@@ -411,90 +283,199 @@ impl MapAlloc {
         // any alignment less than a page
         if new_layout.align() > self.pagesize {
-            return Err(AllocErr::invalid_input("cannot support alignment greater than a page"));
+            return Err(AllocErr::invalid_input(
+                "cannot support alignment greater than a page",
+            ));
         }
-        debug_assert!(layout.size() > 0 && new_layout.size() > 0,
-                      "usable_size: size of layout and new_layout must be non-zero");
 
         let old_size = next_multiple(layout.size(), self.pagesize);
         let new_size = next_multiple(new_layout.size(), self.pagesize);
+
         if old_size == new_size {
             return Ok(ptr);
         }
-        self.realloc_helper(ptr, old_size, new_layout.size()).ok_or(AllocErr::Exhausted { request: new_layout })
+
+        if !cfg!(windows) {
+            // Windows cannot grow or shrink existing mappings.
+            if new_size >= old_size {
+                if let Ok(()) = self.grow_in_place(ptr, layout.clone(), new_layout.clone()) {
+                    return Ok(ptr);
+                }
+            } else if new_size < old_size {
+                if let Ok(()) = self.shrink_in_place(ptr, layout.clone(), new_layout.clone()) {
+                    return Ok(ptr);
+                }
+            }
+        }
+
+        // otherwise, fall back on alloc + copy + dealloc.
+        let result = Alloc::alloc(self, new_layout);
+        if let Ok(new_ptr) = result {
+            use core::cmp;
+
+            // If we're configured to allocate un-readable or un-writable memory, we need to
+            // temporarily mark the old allocation readable and the new allocation writable in
+            // order to be able to copy the data.
+            let fix_permissions = !self.read || !self.write;
+            if fix_permissions {
+                protect(ptr, old_size, perms::get_perm(true, self.write, self.exec));
+                protect(
+                    new_ptr,
+                    new_size,
+                    perms::get_perm(self.read, true, self.exec),
+                );
+            }
+            ptr::copy_nonoverlapping(ptr as *const u8, new_ptr, cmp::min(old_size, new_size));
+            Alloc::dealloc(self, ptr, layout);
+            if fix_permissions {
+                protect(new_ptr, new_size, self.perms)
+            };
+        }
+        result
     }
 
     #[cfg(target_os = "linux")]
-    unsafe fn grow_in_place(&mut self, ptr: *mut u8, layout: Layout, new_layout: Layout) -> Result<(), CannotReallocInPlace> {
+    unsafe fn grow_in_place(
+        &mut self,
+        ptr: *mut u8,
+        layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(), CannotReallocInPlace> {
+        debug_assert!(
+            layout.size() > 0 && new_layout.size() > 0,
+            "grow_in_place: size of layout and new_layout must be non-zero"
+        );
```

Comment:
I think I'd split this into two asserts, but I don't have a strong opinion either way.

Reply:
Fixed.